File: .agent-src/templates/scripts/work_engine/directives/ui/polish.py
"""``polish`` step — bounded fix loop for review findings.

Phase 3 Step 5 of ``agents/roadmaps/road-to-product-ui-track.md``: the
polish step drives the fix loop after ``review`` produces findings.
It is the **only** step that re-enters review during a single
``/work`` run, and the loop is hard-capped at two rounds — anything
the agent cannot fix in two passes goes back to the user as a
ship-as-is / abort decision rather than burning rounds silently.

Routes on ``state.ui_review`` and ``state.ui_polish``:

- **review_clean is True (or findings empty)** — nothing to polish;
  return ``SUCCESS`` so the dispatcher advances to ``report``.
- **review_clean is False, rounds < 2** — emit
  ``@agent-directive: ui-polish-<stack>``. The agent applies the
  fixes from ``state.ui_review.findings``, re-runs the review skill,
  and increments ``state.ui_polish.rounds``. On rebound, the
  dispatcher walks back here; if the new review is clean we succeed,
  otherwise the next round fires.
- **review_clean is False, rounds == 2** — ceiling reached. Halt
  with three numbered options (ship as-is / abort / hand off) so the
  user breaks the deadlock instead of the engine looping forever.

Idempotent: when the review is clean the step round-trips through
``SUCCESS`` regardless of how many rounds ran. Schema-level validation
in :mod:`work_engine.state` already rejects ``rounds > 2`` on disk,
so the ceiling check here is the runtime mirror of that contract.
"""
from __future__ import annotations

from typing import Any

from ...delivery_state import (
    DeliveryState,
    Outcome,
    StepResult,
    agent_directive,
)

POLISH_CEILING = 2
"""Maximum number of polish rounds per ``/work`` run.

Mirrored by :func:`work_engine.state._validate_ui_polish` (rejects
``rounds > 2`` at schema load) so the contract holds across
in-memory state, on-disk state, and the dispatcher.
"""

TOKEN_VIOLATION_KIND = "token_violation"
"""Marker on a review finding that flags a hardcoded design value.

Findings with ``kind == TOKEN_VIOLATION_KIND`` carry ``category``
(``"colors"`` / ``"spacing"`` / ``"typography"`` / …) and ``value``
(the literal hardcoded string). Polish classifies them into matched
(value already present in ``state.ui_audit.design_tokens``) and
unmatched, then reacts per :data:`TOKEN_REPEAT_THRESHOLD`.
"""

TOKEN_REPEAT_THRESHOLD = 2
"""Number of repeats above which an unmatched value triggers the
extraction halt — mirrors the roadmap's ">2 times" wording in
``agents/roadmaps/road-to-product-ui-track.md`` Phase 3 Step 5.
"""

STACK_DIRECTIVES: dict[str, str] = {
    "blade-livewire-flux": "ui-polish-blade-livewire-flux",
    "react-shadcn": "ui-polish-react-shadcn",
    "vue": "ui-polish-vue",
    "plain": "ui-polish-plain",
}
"""Map ``state.stack.frontend`` → agent-directive skill name.

Mirrors :data:`work_engine.directives.ui.review.STACK_DIRECTIVES`.
An unknown stack falls through to ``ui-polish-plain``.
"""

DEFAULT_DIRECTIVE = "ui-polish-plain"
"""Fallback directive when ``state.stack`` is missing or malformed."""

AMBIGUITIES: tuple[dict[str, str], ...] = (
    {
        "code": "polish_round_pending",
        "trigger": "state.ui_review.review_clean is False and "
        "state.ui_polish.rounds < 2 — fixes have not yet been applied "
        "for the current findings",
        "resolution": "agent directive `ui-polish-<stack>` → skill "
        "applies fixes, re-runs the review, increments "
        "state.ui_polish.rounds",
    },
    {
        "code": "polish_ceiling_reached",
        "trigger": "state.ui_polish.rounds == 2 and review still dirty "
        "— two rounds did not converge",
        "resolution": "user picks: ship as-is, abort, or hand off to "
        "manual fix; engine refuses to start a third round",
    },
    {
        "code": "polish_token_extraction_pending",
        "trigger": "state.ui_review.findings has token_violation entries "
        "whose value repeats >2 times and has no match in "
        "state.ui_audit.design_tokens",
        "resolution": "user picks: extract as a new token (agent adds "
        "it to state.ui_audit.design_tokens.<category>), inline (agent "
        "drops the token_violation findings before re-entering polish), "
        "or abort the UI request",
    },
)
"""Declared ambiguity surfaces for this step."""


def run(state: DeliveryState) -> StepResult:
    """Apply the polish-loop gate."""
    review = state.ui_review or {}
    findings = review.get("findings", [])
    if not isinstance(findings, list):
        findings = []
    review_clean = bool(review.get("review_clean", False))

    if review_clean or not findings:
        return StepResult(outcome=Outcome.SUCCESS)

    polish = state.ui_polish or {}
    rounds = polish.get("rounds", 0)
    if not isinstance(rounds, int) or isinstance(rounds, bool):
        rounds = 0

    if rounds >= POLISH_CEILING:
        return _halt_ceiling(state, findings_count=len(findings), rounds=rounds)

    tokens = _design_tokens(state)
    matched, unmatched_repeats = _classify_token_violations(findings, tokens)
    if unmatched_repeats:
        return _halt_token_extraction(state, repeats=unmatched_repeats)

    return _delegate_to_polish_skill(
        state,
        findings_count=len(findings),
        matched_token_count=len(matched),
        rounds=rounds,
    )


def _resolve_directive(state: DeliveryState) -> str:
    """Pick the agent directive for the project's frontend stack."""
    stack = getattr(state, "stack", None) or {}
    if isinstance(stack, dict):
        frontend = stack.get("frontend")
        if isinstance(frontend, str) and frontend in STACK_DIRECTIVES:
            return STACK_DIRECTIVES[frontend]
    return DEFAULT_DIRECTIVE


def _stack_label(state: DeliveryState) -> str:
    """Return the frontend stack label, defaulting to ``plain``."""
    stack = getattr(state, "stack", None) or {}
    if isinstance(stack, dict):
        frontend = stack.get("frontend")
        if isinstance(frontend, str) and frontend:
            return frontend
    return "plain"


def _delegate_to_polish_skill(
    state: DeliveryState,
    *,
    findings_count: int,
    matched_token_count: int,
    rounds: int,
) -> StepResult:
    """BLOCKED halt — emit the stack-specific polish directive.

    The skill applies fixes from ``state.ui_review.findings``,
    re-runs the review, and writes:

    - new ``state.ui_review`` (refreshed findings + ``review_clean``)
    - ``state.ui_polish.rounds = rounds + 1``
    - ``state.ui_polish.applied`` extended with this round's fix log

    ``matched_token_count`` reports how many findings are
    ``token_violation`` entries whose value already lives in
    ``state.ui_audit.design_tokens`` — those auto-convert in this
    round without further user input.
    """
    directive = _resolve_directive(state)
    stack_label = _stack_label(state)
    next_round = rounds + 1
    findings_line = (
        f"> {findings_count} finding(s) from `state.ui_review`. "
        "Apply each fix, re-run the review, and write the refreshed "
        "envelope back."
    )
    if matched_token_count:
        findings_line = (
            f"> {findings_count} finding(s) from `state.ui_review` "
            f"({matched_token_count} token-violation match(es) "
            "auto-convert against `state.ui_audit.design_tokens`). "
            "Apply each fix, re-run the review, and write the refreshed "
            "envelope back."
        )
    return StepResult(
        outcome=Outcome.BLOCKED,
        questions=[
            agent_directive(directive),
            f"> Stack: `{stack_label}`. Polish round "
            f"{next_round} of {POLISH_CEILING}.",
            findings_line,
            "> 1. Continue \u2014 apply fixes, re-review, and increment "
            "`state.ui_polish.rounds`",
            "> 2. Abort \u2014 drop this UI request",
        ],
        message=(
            f"UI polish round {next_round}/{POLISH_CEILING}; delegating "
            f"to `{directive}` for stack `{stack_label}`."
        ),
    )


def _halt_ceiling(
    state: DeliveryState,
    *,
    findings_count: int,
    rounds: int,
) -> StepResult:
    """BLOCKED halt — two rounds spent and review still dirty."""
    stack_label = _stack_label(state)
    return StepResult(
        outcome=Outcome.BLOCKED,
        questions=[
            f"> Stack: `{stack_label}`. Polish ceiling reached "
            f"({rounds}/{POLISH_CEILING} rounds).",
            f"> {findings_count} finding(s) still open in "
            "`state.ui_review`. The engine refuses a third round.",
            "> 1. Ship as-is \u2014 mark `state.ui_review.review_clean "
            "= True` and continue to `report` (the open findings stay "
            "in the delivery report)",
            "> 2. Abort \u2014 drop this UI request",
            "> 3. Hand off \u2014 a human picks up the remaining "
            "findings outside the engine; re-invoke `/work` only after "
            "they are resolved",
            "",
            "**Recommendation: 3 \u2014 Hand off** \u2014 two automated "
            "rounds failed to converge. Caveat: pick 1 only when the "
            "remaining findings are explicitly acceptable (low-priority "
            "polish, deferred to a follow-up).",
        ],
        message=(
            f"UI polish ceiling reached ({rounds}/{POLISH_CEILING}); "
            f"{findings_count} finding(s) still open."
        ),
    )


def _design_tokens(state: DeliveryState) -> dict[str, Any]:
    """Return ``state.ui_audit.design_tokens`` as a dict, or ``{}``.

    Defensive: tolerates missing audit, non-dict audit, missing
    ``design_tokens`` key, and non-dict ``design_tokens`` payload.
    All four shapes degrade to "no tokens known" so the unmatched
    classifier treats every value as an extraction candidate.
    """
    audit = getattr(state, "ui_audit", None) or {}
    if not isinstance(audit, dict):
        return {}
    tokens = audit.get("design_tokens") or {}
    if not isinstance(tokens, dict):
        return {}
    return tokens


def _classify_token_violations(
    findings: list[Any],
    tokens: dict[str, Any],
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    """Split ``token_violation`` findings into matched / unmatched-repeats.

    A finding qualifies when ``kind == TOKEN_VIOLATION_KIND`` and it
    carries string ``category`` + ``value``. Matched: ``value`` is
    among the values of ``tokens[category]`` (the token bucket is a
    name → literal mapping per ``state.ui_audit.design_tokens``
    schema). Unmatched values are bucketed by ``(category, value)``;
    only buckets with ``count > TOKEN_REPEAT_THRESHOLD`` are returned
    so single-use hardcoded values do not trigger the halt.
    """
    matched: list[dict[str, Any]] = []
    unmatched_counts: dict[tuple[str, str], int] = {}
    for finding in findings:
        if not isinstance(finding, dict):
            continue
        if finding.get("kind") != TOKEN_VIOLATION_KIND:
            continue
        category = finding.get("category")
        value = finding.get("value")
        if not isinstance(category, str) or not isinstance(value, str):
            continue
        bucket = tokens.get(category)
        if isinstance(bucket, dict) and value in bucket.values():
            matched.append(finding)
            continue
        key = (category, value)
        unmatched_counts[key] = unmatched_counts.get(key, 0) + 1
    repeats = [
        {"category": cat, "value": val, "count": count}
        for (cat, val), count in unmatched_counts.items()
        if count > TOKEN_REPEAT_THRESHOLD
    ]
    return matched, repeats


def _suggest_token_name(category: str, value: str) -> str:
    """Build a suggested CSS-custom-property name for an extraction halt.

    Heuristic only — the agent picks the final name when applying
    the extraction. Strips non-alphanumerics from ``value``, prefixes
    with the singular of ``category`` (``colors`` → ``color``), and
    caps the length so the halt body stays readable.
    """
    safe = "".join(c if c.isalnum() else "-" for c in value).strip("-").lower()
    if not safe:
        safe = "value"
    base = category.rstrip("s") or category
    return f"{base}-{safe}"[:40]


def _halt_token_extraction(
    state: DeliveryState,
    *,
    repeats: list[dict[str, Any]],
) -> StepResult:
    """BLOCKED halt — repeated hardcoded value(s) without a matching token.

    Polish refuses to silently inline the same hardcoded value across
    multiple call sites; the user picks whether the value graduates
    to a design token or stays inline for this run.
    """
    stack_label = _stack_label(state)
    questions: list[str] = [
        f"> Stack: `{stack_label}`. {len(repeats)} hardcoded value(s) "
        f"appear >{TOKEN_REPEAT_THRESHOLD} times without a matching "
        "entry in `state.ui_audit.design_tokens`.",
    ]
    for repeat in repeats:
        suggested = _suggest_token_name(repeat["category"], repeat["value"])
        questions.append(
            f"> - `{repeat['value']}` "
            f"({repeat['category']}, {repeat['count']}\u00d7) "
            f"\u2014 suggested name: `--{suggested}`"
        )
    questions.extend([
        "> 1. Extract as design token(s) \u2014 add to "
        "`state.ui_audit.design_tokens.<category>` and re-enter polish; "
        "matching findings auto-convert next round",
        "> 2. Inline \u2014 keep the hardcoded value(s) for this run; "
        "drop the token_violation findings from "
        "`state.ui_review.findings` before re-entering polish",
        "> 3. Abort \u2014 drop this UI request",
        "",
        "**Recommendation: 1 \u2014 Extract** \u2014 a value used "
        f">{TOKEN_REPEAT_THRESHOLD} times is a de-facto token; "
        "promoting it now keeps the design system honest. Pick 2 only "
        "when the value is intentionally one-off.",
    ])
    return StepResult(
        outcome=Outcome.BLOCKED,
        questions=questions,
        message=(
            f"UI polish paused; {len(repeats)} hardcoded value(s) "
            "repeat without a matching design token."
        ),
    )


__all__ = [
    "AMBIGUITIES",
    "DEFAULT_DIRECTIVE",
    "POLISH_CEILING",
    "STACK_DIRECTIVES",
    "TOKEN_REPEAT_THRESHOLD",
    "TOKEN_VIOLATION_KIND",
    "run",
]

File: src/agent-src/templates/scripts/work_engine/directives/ui/polish.ts
/**
 * `polish` step — bounded fix loop for review findings.
 *
 * TypeScript twin of `directives/ui/polish.py` (ADR-094 py2ts). Public API
 * names stay snake_case to mirror the Python module 1:1.
 *
 * The polish step drives the fix loop after `review` produces findings. The
 * loop is hard-capped at two rounds (extendable by one for a11y); anything the
 * agent cannot fix goes back to the user as a ship-as-is / abort decision.
 */
import {
    type Any,
    type DeliveryState,
    Outcome,
    StepResult,
    agent_directive,
} from '../../delivery_state.js';

/** Maximum number of polish rounds per `/work` run. */
export const POLISH_CEILING = 2;

/** Marker on a review finding that flags an a11y issue. */
export const A11Y_VIOLATION_KIND = 'a11y_violation';

/** Marker on a review finding that flags a hardcoded design value. */
export const TOKEN_VIOLATION_KIND = 'token_violation';

/** Repeat count above which an unmatched value triggers the extraction halt. */
export const TOKEN_REPEAT_THRESHOLD = 2;

/** Map `state.stack.frontend` → agent-directive skill name. */
export const STACK_DIRECTIVES: Record<string, string> = {
    'blade-livewire-flux': 'ui-polish-blade-livewire-flux',
    'react-shadcn': 'ui-polish-react-shadcn',
    vue: 'ui-polish-vue',
    plain: 'ui-polish-plain',
};

/** Fallback directive when `state.stack` is missing or malformed. */
export const DEFAULT_DIRECTIVE = 'ui-polish-plain';

export const AMBIGUITIES: ReadonlyArray<Record<string, string>> = [
    {
        code: 'polish_round_pending',
        trigger:
            'state.ui_review.review_clean is False and ' +
            'state.ui_polish.rounds < 2 — fixes have not yet been applied ' +
            'for the current findings',
        resolution:
            'agent directive `ui-polish-<stack>` → skill ' +
            'applies fixes, re-runs the review, increments ' +
            'state.ui_polish.rounds',
    },
    {
        code: 'polish_ceiling_reached',
        trigger:
            'state.ui_polish.rounds == ceiling and remaining ' +
            'findings are non-a11y (subjective polish that did not ' +
            'converge) — a11y blocks take precedence via ' +
            'polish_a11y_blocking',
        resolution:
            'user picks: ship as-is, abort, or hand off to ' +
            'manual fix; engine refuses to start another round',
    },
    {
        code: 'polish_a11y_blocking',
        trigger:
            'state.ui_polish.rounds == ceiling and ' +
            'state.ui_review.findings still contains a11y_violation ' +
            'entries — objective gate that takes precedence over the ' +
            'subjective polish_ceiling_reached halt',
        resolution:
            'user picks: extend by one round (engine sets ' +
            'state.ui_polish.extension_used=True so the next round can ' +
            'fire), accept-with-known-violations (engine appends the ' +
            'leftover violations to state.ui_review.a11y.accepted_violations ' +
            'so the review gate stops blocking on them), or abort the ' +
            'UI request',
    },
    {
        code: 'polish_token_extraction_pending',
        trigger:
            'state.ui_review.findings has token_violation entries ' +
            'whose value repeats >2 times and has no match in ' +
            'state.ui_audit.design_tokens',
        resolution:
            'user picks: extract as a new token (agent adds ' +
            'it to state.ui_audit.design_tokens.<category>), inline (agent ' +
            'drops the token_violation findings before re-entering polish), ' +
            'or abort the UI request',
    },
];

function _isDict(value: Any): value is Record<string, Any> {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function _pyTruthy(value: Any): boolean {
    if (value === null || value === undefined) return false;
    if (typeof value === 'string') return value.length > 0;
    if (typeof value === 'number') return value !== 0;
    if (typeof value === 'boolean') return value;
    if (Array.isArray(value)) return value.length > 0;
    if (typeof value === 'object') return Object.keys(value).length > 0;
    return true;
}

/** Apply the polish-loop gate. */
export function run(state: DeliveryState): StepResult {
    const review = _pyTruthy(state.ui_review) ? (state.ui_review as Record<string, Any>) : {};
    let findings = 'findings' in review ? review['findings'] : [];
    if (!Array.isArray(findings)) {
        findings = [];
    }
    const review_clean = _pyTruthy(review['review_clean']);

    if (review_clean || (findings as Any[]).length === 0) {
        return new StepResult({ outcome: Outcome.SUCCESS });
    }

    const polish = _pyTruthy(state.ui_polish) ? (state.ui_polish as Record<string, Any>) : {};
    let rounds = 'rounds' in polish ? polish['rounds'] : 0;
    // `not isinstance(rounds, int) or isinstance(rounds, bool)` → reset to 0.
    if (typeof rounds !== 'number' || !Number.isInteger(rounds) || typeof rounds === 'boolean') {
        rounds = 0;
    }
    const extension_used = _pyTruthy(polish['extension_used']);
    const effective_ceiling = POLISH_CEILING + (extension_used ? 1 : 0);

    const findingsArr = findings as Any[];
    if ((rounds as number) >= effective_ceiling) {
        const a11y_findings = _a11y_findings(findingsArr);
        if (a11y_findings.length > 0) {
            return _halt_a11y_blocking(state, a11y_findings, rounds as number, !extension_used);
        }
        return _halt_ceiling(state, findingsArr.length, rounds as number, effective_ceiling);
    }

    const tokens = _design_tokens(state);
    const [matched, unmatched_repeats] = _classify_token_violations(findingsArr, tokens);
    if (unmatched_repeats.length > 0) {
        return _halt_token_extraction(state, unmatched_repeats);
    }

    return _delegate_to_polish_skill(
        state,
        findingsArr.length,
        matched.length,
        rounds as number,
        effective_ceiling,
    );
}

/** Pick the agent directive for the project's frontend stack. */
function _resolve_directive(state: DeliveryState): string {
    const stack = _pyTruthy(state.stack) ? state.stack : {};
    if (_isDict(stack)) {
        const frontend = stack['frontend'];
        if (typeof frontend === 'string' && frontend in STACK_DIRECTIVES) {
            return STACK_DIRECTIVES[frontend] as string;
        }
    }
    return DEFAULT_DIRECTIVE;
}

/** Return the frontend stack label, defaulting to `plain`. */
function _stack_label(state: DeliveryState): string {
    const stack = _pyTruthy(state.stack) ? state.stack : {};
    if (_isDict(stack)) {
        const frontend = stack['frontend'];
        if (typeof frontend === 'string' && frontend !== '') {
            return frontend;
        }
    }
    return 'plain';
}

/** BLOCKED halt — emit the stack-specific polish directive. */
function _delegate_to_polish_skill(
    state: DeliveryState,
    findings_count: number,
    matched_token_count: number,
    rounds: number,
    ceiling: number,
): StepResult {
    const directive = _resolve_directive(state);
    const stack_label = _stack_label(state);
    const next_round = rounds + 1;
    let findings_line =
        `> ${findings_count} finding(s) from \`state.ui_review\`. ` +
        'Apply each fix, re-run the review, and write the refreshed ' +
        'envelope back.';
    if (matched_token_count) {
        findings_line =
            `> ${findings_count} finding(s) from \`state.ui_review\` ` +
            `(${matched_token_count} token-violation match(es) ` +
            'auto-convert against `state.ui_audit.design_tokens`). ' +
            'Apply each fix, re-run the review, and write the refreshed ' +
            'envelope back.';
    }
    return new StepResult({
        outcome: Outcome.BLOCKED,
        questions: [
            agent_directive(directive),
            `> Stack: \`${stack_label}\`. Polish round ${next_round} of ${ceiling}.`,
            findings_line,
            '> Fix chart-type / contrast findings against the adopted ' +
                "corpus rows (design-intelligence § 'Grounding the " +
                "review/polish a11y gate'), not ad-hoc judgment.",
            '> 1. Continue — apply fixes, re-review, and increment ' +
                '`state.ui_polish.rounds`',
            '> 2. Abort — drop this UI request',
        ],
        message:
            `UI polish round ${next_round}/${ceiling}; delegating ` +
            `to \`${directive}\` for stack \`${stack_label}\`.`,
    });
}

/** BLOCKED halt — ceiling reached on subjective (non-a11y) findings. */
function _halt_ceiling(
    state: DeliveryState,
    findings_count: number,
    rounds: number,
    ceiling: number,
): StepResult {
    const stack_label = _stack_label(state);
    return new StepResult({
        outcome: Outcome.BLOCKED,
        questions: [
            `> Stack: \`${stack_label}\`. Polish ceiling reached ` +
                `(${rounds}/${ceiling} rounds).`,
            `> ${findings_count} finding(s) still open in ` +
                '`state.ui_review`. The engine refuses a third round.',
            '> 1. Ship as-is — mark `state.ui_review.review_clean ' +
                '= True` and continue to `report` (the open findings stay ' +
                'in the delivery report)',
            '> 2. Abort — drop this UI request',
            '> 3. Hand off — a human picks up the remaining ' +
                'findings outside the engine; re-invoke `/work` only after ' +
                'they are resolved',
            '',
            '**Recommendation: 3 — Hand off** — two automated ' +
                'rounds failed to converge. Caveat: pick 1 only when the ' +
                'remaining findings are explicitly acceptable (low-priority ' +
                'polish, deferred to a follow-up).',
        ],
        message:
            `UI polish ceiling reached (${rounds}/${ceiling}); ` +
            `${findings_count} finding(s) still open.`,
    });
}

/** Return the subset of `findings` synthesised by the a11y gate. */
function _a11y_findings(findings: Any[]): Record<string, Any>[] {
    return findings.filter(
        (f): f is Record<string, Any> => _isDict(f) && f['kind'] === A11Y_VIOLATION_KIND,
    );
}

/** BLOCKED halt — ceiling reached with actionable a11y findings. */
function _halt_a11y_blocking(
    state: DeliveryState,
    a11y_findings: Record<string, Any>[],
    rounds: number,
    extension_available: boolean,
): StepResult {
    const stack_label = _stack_label(state);
    const count = a11y_findings.length;
    const questions: string[] = [
        `> Stack: \`${stack_label}\`. Polish ceiling reached ` +
            `(${rounds}/${POLISH_CEILING} rounds) with ${count} a11y ` +
            'violation(s) still open.',
    ];
    for (const finding of a11y_findings.slice(0, 5)) {
        const rule = _pyTruthy(finding['rule']) ? finding['rule'] : '?';
        const selector = _pyTruthy(finding['selector']) ? finding['selector'] : '?';
        const severity = _pyTruthy(finding['severity']) ? finding['severity'] : '?';
        questions.push(`> - \`${pyStr(rule)}\` on \`${pyStr(selector)}\` (severity: ${pyStr(severity)})`);
    }
    if (count > 5) {
        questions.push(`> ... and ${count - 5} more`);
    }
    if (extension_available) {
        questions.push(
            '> 1. Extend — grant one extra polish round; the ' +
                'engine sets `state.ui_polish.extension_used = True` so ' +
                'the next delegation can fire',
            '> 2. Accept — append the open violations to ' +
                '`state.ui_review.a11y.accepted_violations` so the review ' +
                'gate stops blocking on them, then continue to `report`',
            '> 3. Abort — drop this UI request',
            '',
            '**Recommendation: 1 — Extend** — a11y ' +
                'violations are objective; one more round usually closes ' +
                'the gap. Pick 2 only when the violations are explicitly ' +
                'out of scope for this run.',
        );
    } else {
        questions.push(
            '> 1. Accept — append the open violations to ' +
                '`state.ui_review.a11y.accepted_violations` so the review ' +
                'gate stops blocking on them, then continue to `report`',
            '> 2. Abort — drop this UI request',
            '',
            '**Recommendation: 1 — Accept** — the one-shot ' +
                'extension is already spent; either accept the residual ' +
                'violations or abort.',
        );
    }
    return new StepResult({
        outcome: Outcome.BLOCKED,
        questions,
        message:
            `UI polish ceiling reached (${rounds}/${POLISH_CEILING}); ` +
            `${count} a11y violation(s) still open.`,
    });
}

/** Return `state.ui_audit.design_tokens` as a dict, or `{}`. */
function _design_tokens(state: DeliveryState): Record<string, Any> {
    const audit = _pyTruthy(state.ui_audit) ? state.ui_audit : {};
    if (!_isDict(audit)) {
        return {};
    }
    const tokens = _pyTruthy(audit['design_tokens']) ? audit['design_tokens'] : {};
    if (!_isDict(tokens)) {
        return {};
    }
    return tokens;
}

/** Split `token_violation` findings into matched / unmatched-repeats. */
function _classify_token_violations(
    findings: Any[],
    tokens: Record<string, Any>,
): [Record<string, Any>[], Record<string, Any>[]] {
    const matched: Record<string, Any>[] = [];
    // Insertion-ordered map keyed by `${category} ${value}` to mirror the
    // Python `dict[tuple[str, str], int]` insertion order.
    const unmatched_counts = new Map<string, { category: string; value: string; count: number }>();
    for (const finding of findings) {
        if (!_isDict(finding)) {
            continue;
        }
        if (finding['kind'] !== TOKEN_VIOLATION_KIND) {
            continue;
        }
        const category = finding['category'];
        const value = finding['value'];
        if (typeof category !== 'string' || typeof value !== 'string') {
            continue;
        }
        const bucket = tokens[category];
        if (_isDict(bucket) && Object.values(bucket).includes(value)) {
            matched.push(finding);
            continue;
        }
        const key = `${category} ${value}`;
        const existing = unmatched_counts.get(key);
        if (existing) {
            existing.count += 1;
        } else {
            unmatched_counts.set(key, { category, value, count: 1 });
        }
    }
    const repeats: Record<string, Any>[] = [];
    for (const { category, value, count } of unmatched_counts.values()) {
        if (count > TOKEN_REPEAT_THRESHOLD) {
            repeats.push({ category, value, count });
        }
    }
    return [matched, repeats];
}

/** Build a suggested CSS-custom-property name for an extraction halt. */
function _suggest_token_name(category: string, value: string): string {
    let safe = '';
    for (const c of value) {
        safe += _isAlnum(c) ? c : '-';
    }
    safe = _pyStripChar(safe, '-').toLowerCase();
    if (safe === '') {
        safe = 'value';
    }
    const base = _pyRStripChar(category, 's') || category;
    return [...`${base}-${safe}`].slice(0, 40).join('');
}

/** BLOCKED halt — repeated hardcoded value(s) without a matching token. */
function _halt_token_extraction(
    state: DeliveryState,
    repeats: Record<string, Any>[],
): StepResult {
    const stack_label = _stack_label(state);
    const questions: string[] = [
        `> Stack: \`${stack_label}\`. ${repeats.length} hardcoded value(s) ` +
            `appear >${TOKEN_REPEAT_THRESHOLD} times without a matching ` +
            'entry in `state.ui_audit.design_tokens`.',
    ];
    for (const repeat of repeats) {
        const suggested = _suggest_token_name(repeat['category'] as string, repeat['value'] as string);
        questions.push(
            `> - \`${pyStr(repeat['value'])}\` ` +
                `(${pyStr(repeat['category'])}, ${pyStr(repeat['count'])}×) ` +
                `— suggested name: \`--${suggested}\``,
        );
    }
    questions.push(
        '> 1. Extract as design token(s) — add to ' +
            '`state.ui_audit.design_tokens.<category>` and re-enter polish; ' +
            'matching findings auto-convert next round',
        '> 2. Inline — keep the hardcoded value(s) for this run; ' +
            'drop the token_violation findings from ' +
            '`state.ui_review.findings` before re-entering polish',
        '> 3. Abort — drop this UI request',
        '',
        '**Recommendation: 1 — Extract** — a value used ' +
            `>${TOKEN_REPEAT_THRESHOLD} times is a de-facto token; ` +
            'promoting it now keeps the design system honest. Pick 2 only ' +
            'when the value is intentionally one-off.',
    );
    return new StepResult({
        outcome: Outcome.BLOCKED,
        questions,
        message:
            `UI polish paused; ${repeats.length} hardcoded value(s) ` +
            'repeat without a matching design token.',
    });
}

// ── Python string helpers ───────────────────────────────────────────────

function pyStr(value: Any): string {
    if (value === null || value === undefined) return 'None';
    if (value === true) return 'True';
    if (value === false) return 'False';
    return String(value);
}

/** Python `str.isalnum()` for a single code point (Unicode-aware). */
function _isAlnum(ch: string): boolean {
    // Python: alphanumeric = letters (incl. Unicode) or any numeric character.
    // \p{L} (letters), \p{N} (numbers) cover str.isalnum for single chars.
    return /[\p{L}\p{N}]/u.test(ch);
}

/** Python `str.strip(chars)` — strip the given chars from both ends. */
function _pyStripChar(s: string, ch: string): string {
    let start = 0;
    let end = s.length;
    while (start < end && s[start] === ch) start += 1;
    while (end > start && s[end - 1] === ch) end -= 1;
    return s.slice(start, end);
}

/** Python `str.rstrip(chars)` — strip the given chars from the right. */
function _pyRStripChar(s: string, ch: string): string {
    let end = s.length;
    while (end > 0 && s[end - 1] === ch) end -= 1;
    return s.slice(0, end);
}
